File: services/cobalt_client.py
import logging
import aiohttp
from urllib.parse import urlparse
from pathlib import Path
from typing import Optional, Dict, List, Tuple, Callable
from config import COBALT_API_URL, COBALT_API_KEY, HTTP_PROXY, HTTPS_PROXY, DOWNLOADS_DIR


class CobaltClient:
    """
    Клиент для работы с Cobalt API.
    Прокси настраивается только для запросов к Cobalt, не влияет на бота.
    """
    
    def __init__(self):
        self.api_url = COBALT_API_URL.rstrip('/')
        self.api_key = COBALT_API_KEY
        self.proxy = None  # Прокси не используется - Cobalt воркер сам обращается к сайтам
        
        logging.info(f"Cobalt client: using external API at {self.api_url}")
    
    def _mask_proxy(self, proxy_url: str) -> str:
        """Маскирует credentials в proxy URL для логирования"""
        try:
            if '@' in proxy_url:
                protocol, rest = proxy_url.split('://', 1)
                credentials, host = rest.split('@', 1)
                return f"{protocol}://***:***@{host}"
            return proxy_url
        except:
            return "***"
    
    async def _make_request(self, url: str, method: str = "GET", json_data: Dict = None) -> Dict:
        """
        Выполняет HTTP запрос к Cobalt API через прокси (если настроен).
        """
        headers = {
            "Accept": "application/json",
            "Content-Type": "application/json",
        }
        
        if self.api_key:
            headers["Authorization"] = f"Api-Key {self.api_key}"
        
        # Создаем connector с поддержкой прокси
        connector = aiohttp.TCPConnector(ssl=False)
        timeout = aiohttp.ClientTimeout(total=300)  # 5 минут для больших файлов
        
        try:
            async with aiohttp.ClientSession(
                connector=connector, 
                timeout=timeout
            ) as session:
                if method == "POST":
                    async with session.post(
                        url, 
                        json=json_data, 
                        headers=headers,
                        proxy=self.proxy  # Прокси только для Cobalt запросов
                    ) as response:
                        if response.status == 200:
                            return await response.json()
                        else:
                            error_text = await response.text()
                            logging.error(f"Cobalt API error {response.status}: {error_text}")
                            raise Exception(f"Cobalt API returned {response.status}: {error_text}")
                else:
                    async with session.get(
                        url, 
                        headers=headers,
                        proxy=self.proxy
                    ) as response:
                        if response.status == 200:
                            return await response.json()
                        else:
                            error_text = await response.text()
                            logging.error(f"Cobalt API error {response.status}: {error_text}")
                            raise Exception(f"Cobalt API returned {response.status}")
        except Exception as e:
            logging.error(f"Request to Cobalt failed: {str(e)}")
            raise
    
    async def download_media(
        self, 
        url: str, 
        quality: str = "1080",
        is_audio: bool = False,
        progress_callback: Optional[Callable] = None
    ) -> Tuple[Optional[Path], Optional[Path], Dict]:
        """
        Загружает медиа через Cobalt API.
        
        Args:
            url: URL видео/аудио
            quality: Качество видео (max, 2160, 1440, 1080, 720, 480, 360, 240, 144)
            is_audio: True если нужно скачать только аудио
            progress_callback: Callback для обновления прогресса
        
        Returns:
            (file_path, thumbnail_path, metadata)
        """
        logging.info(f"Cobalt: downloading {url} (quality={quality}, audio={is_audio})")
        
        request_data = {
            "url": url,
            "videoQuality": quality,
            "youtubeVideoCodec": "h264",  # Всегда запрашиваем H.264 для Telegram
            "downloadMode": "audio" if is_audio else "auto",
            "alwaysProxy": False  # Получаем прямые ссылки, не туннель
        }
        
        if is_audio:
            request_data["audioFormat"] = "mp3"
            request_data["audioBitrate"] = "320"
        
        try:
            response = await self._make_request(
                f"{self.api_url}/",
                method="POST",
                json_data=request_data
            )
            
            status = response.get("status")
            logging.info(f"Cobalt response status: {status}")
            
            if status == "error":
                error_code = response.get("error", {}).get("code", "unknown")
                error_msg = response.get("error", {}).get("context", "Unknown error")
                logging.error(f"Cobalt error: {error_code} - {error_msg}")
                raise Exception(f"Cobalt error: {error_msg}")
            
            elif status == "redirect":
                # Прямая ссылка на файл
                file_url = response.get("url")
                return await self._download_file(file_url, response, progress_callback)
            
            elif status == "tunnel":
                # Cobalt проксирует файл через /tunnel
                tunnel_url = response.get("url")
                return await self._download_file(tunnel_url, response, progress_callback)
            
            elif status == "picker":
                # Множественные файлы (например, TikTok слайдшоу)
                return await self._handle_picker(response)
            
            elif status == "local-processing":
                # Требуется локальная обработка (merge audio+video)
                logging.warning("Cobalt returned local-processing, needs manual FFmpeg merge")
                raise Exception("Local processing not implemented yet, use fallback")
            
            else:
                raise Exception(f"Unknown Cobalt status: {status}")
        
        except Exception as e:
            logging.error(f"Cobalt download failed: {str(e)}")
            raise
    
    async def _download_file(
        self, 
        file_url: str, 
        metadata: Dict,
        progress_callback: Optional[Callable] = None
    ) -> Tuple[Path, Optional[Path], Dict]:
        """
        Скачивает файл по URL (redirect или tunnel) с поддержкой прогресса.
        """
        import uuid
        
        filename = metadata.get("filename", f"video_{uuid.uuid4().hex[:8]}.mp4")
        # Очистка имени файла от недопустимых символов
        filename = "".join(c for c in filename if c.isalnum() or c in "._- ")
        
        file_path = DOWNLOADS_DIR / filename
        
        logging.info(f"Downloading file from: {file_url}")
        
        connector = aiohttp.TCPConnector(ssl=False)
        timeout = aiohttp.ClientTimeout(total=600)  # 10 минут
        
        try:
            async with aiohttp.ClientSession(
                connector=connector, 
                timeout=timeout
            ) as session:
                async with session.get(file_url, proxy=self.proxy) as response:
                    if response.status != 200:
                        raise Exception(f"Failed to download file: HTTP {response.status}")
                    
                    # Получаем размер файла для прогресса
                    total_size = int(response.headers.get('content-length', 0))
                    downloaded = 0
                    
                    # Скачиваем файл chunked с прогрессом
                    with open(file_path, 'wb') as f:
                        async for chunk in response.content.iter_chunked(8192):
                            f.write(chunk)
                            downloaded += len(chunk)
                            
                            # Обновляем прогресс
                            # Keep download status message stable (funny status only)
            
            if not file_path.exists() or file_path.stat().st_size == 0:
                raise Exception("Downloaded file is empty or doesn't exist")
            
            logging.info(f"File downloaded: {file_path} ({file_path.stat().st_size} bytes)")
            
            # Формируем метаданные
            uploader = metadata.get("author") or metadata.get("uploader") or metadata.get("creator") or "Unknown"
            title = metadata.get("title") or filename.rsplit('.', 1)[0]
            result_metadata = {
                "title": title,
                "uploader": uploader,
                "webpage_url": metadata.get("url", ""),
                "duration": 0,
                "width": 0,
                "height": 0,
                "codec": "h264"  # Cobalt отдает H.264
            }
            
            return file_path, None, result_metadata
        
        except Exception as e:
            if file_path.exists():
                file_path.unlink()
            logging.error(f"File download failed: {str(e)}")
            raise
    
    async def _handle_picker(self, response: Dict) -> Tuple[List[Path], Optional[Path], Dict]:
        """
        Обрабатывает picker response (множественные файлы).
        Например, TikTok слайдшоу.
        """
        logging.info("Cobalt returned picker (multiple files)")
        
        picker_items = response.get("picker", [])
        if not picker_items:
            raise Exception("Picker response is empty")
        
        file_paths = []

        def guess_extension(item_url: str, mime: Optional[str]) -> str:
            if item_url:
                path = urlparse(item_url).path
                if "." in path:
                    ext = Path(path).suffix.lower()
                    if ext:
                        return ext
            if mime:
                mime_lower = mime.lower()
                if "jpeg" in mime_lower:
                    return ".jpg"
                if "png" in mime_lower:
                    return ".png"
                if "webp" in mime_lower:
                    return ".webp"
                if "mp3" in mime_lower:
                    return ".mp3"
                if "m4a" in mime_lower:
                    return ".m4a"
                if "aac" in mime_lower:
                    return ".aac"
                if "ogg" in mime_lower:
                    return ".ogg"
                if "mp4" in mime_lower:
                    return ".mp4"
            return ".bin"

        for idx, item in enumerate(picker_items):
            item_url = item.get("url")
            if not item_url:
                continue

            mime = item.get("mime") or item.get("type")
            ext = guess_extension(item_url, mime)
            filename = f"picker_{idx}{ext}"

            try:
                file_path, _, _ = await self._download_file(
                    item_url,
                    {"filename": filename}
                )
                file_paths.append(file_path)
            except Exception as e:
                logging.error(f"Failed to download picker item {idx}: {e}")

        metadata = {
            "title": response.get("title") or "Carousel Media",
            "uploader": response.get("author") or response.get("uploader") or response.get("creator") or "Unknown",
            "webpage_url": response.get("url", ""),
            "count": len(file_paths)
        }
        
        return file_paths, None, metadata


# Singleton instance
cobalt = CobaltClient()

File: java/src/main/java/me/datapeice/ytttins/service/CobaltClient.java
package me.datapeice.ytttins.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import me.datapeice.ytttins.config.BotConfig;
import okhttp3.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.io.*;
import java.net.URI;
import java.nio.file.*;
import java.util.*;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * HTTP client for the <a href="https://github.com/imputnet/cobalt">Cobalt API</a>.
 *
 * <p>Equivalent to {@code services/cobalt_client.py} in the Python version.
 *
 * <h3>Cobalt API response types handled</h3>
 * <ul>
 *   <li>{@code redirect} – direct CDN link</li>
 *   <li>{@code tunnel}   – proxied through Cobalt's /tunnel endpoint</li>
 *   <li>{@code picker}   – multiple files (e.g. TikTok slideshow)</li>
 *   <li>{@code error}    – Cobalt returned an error</li>
 * </ul>
 */
@Service
public class CobaltClient {

    private static final Logger log = LoggerFactory.getLogger(CobaltClient.class);
    private static final MediaType JSON = MediaType.parse("application/json; charset=utf-8");

    private final BotConfig config;
    private final OkHttpClient http;
    private final ObjectMapper mapper = new ObjectMapper();

    public CobaltClient(BotConfig config) {
        this.config = config;
        this.http = new OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(600, TimeUnit.SECONDS)    // 10 minutes for large files
                .writeTimeout(60, TimeUnit.SECONDS)
                .followRedirects(true)
                .build();
    }

    // ─── Public API ───────────────────────────────────────────────────────────────

    /**
     * Downloads media via the Cobalt API.
     *
     * @param url              source media URL
     * @param quality          requested quality ("max", "2160", "1440", "1080", etc.)
     * @param isAudio          {@code true} to download audio-only (MP3 @ 320 kbps)
     * @param progressCallback called with status updates; may be null
     * @return download result with file path(s) and metadata
     */
    public DownloaderService.DownloadResult downloadMedia(
            String url,
            String quality,
            boolean isAudio,
            Consumer<String> progressCallback) throws Exception {

        if (!config.isCobaltEnabled() || config.getCobaltApiUrl().isBlank()) {
            throw new IllegalStateException("Cobalt API is not configured");
        }

        log.info("Cobalt: downloading {} (quality={}, audio={})", url, quality, isAudio);

        Map<String, Object> requestBody = new LinkedHashMap<>();
        requestBody.put("url", url);
        requestBody.put("videoQuality", quality);
        requestBody.put("youtubeVideoCodec", "h264");
        requestBody.put("downloadMode", isAudio ? "audio" : "auto");
        requestBody.put("alwaysProxy", false);
        if (isAudio) {
            requestBody.put("audioFormat", "mp3");
            requestBody.put("audioBitrate", "320");
        }

        String cobaltBase = config.getCobaltApiUrl();
        while (cobaltBase.endsWith("/")) cobaltBase = cobaltBase.substring(0, cobaltBase.length() - 1);
        JsonNode response = postJson(cobaltBase + "/", requestBody);
        String status = response.path("status").asText();
        log.info("Cobalt response status: {}", status);

        return switch (status) {
            case "redirect", "tunnel" -> {
                String fileUrl = response.path("url").asText();
                yield downloadFile(fileUrl, response, progressCallback, url);
            }
            case "picker" -> handlePicker(response);
            case "error" -> {
                String code = response.path("error").path("code").asText("unknown");
                String msg = response.path("error").path("context").asText("Unknown error");
                throw new RuntimeException("Cobalt error: " + code + " – " + msg);
            }
            default -> throw new RuntimeException("Unknown Cobalt status: " + status);
        };
    }

    // ─── JSON request ─────────────────────────────────────────────────────────────

    private JsonNode postJson(String url, Map<String, Object> body) throws IOException {
        String json = mapper.writeValueAsString(body);
        Request.Builder builder = new Request.Builder()
                .url(url)
                .post(RequestBody.create(json, JSON))
                .header("Accept", "application/json")
                .header("Content-Type", "application/json");

        if (config.getCobaltApiKey() != null && !config.getCobaltApiKey().isBlank()) {
            builder.header("Authorization", "Api-Key " + config.getCobaltApiKey());
        }

        try (Response resp = http.newCall(builder.build()).execute()) {
            String responseBody = resp.body() != null ? resp.body().string() : "";
            if (!resp.isSuccessful()) {
                throw new IOException("Cobalt API error " + resp.code() + ": " + responseBody);
            }
            return mapper.readTree(responseBody);
        }
    }

    // ─── File download ────────────────────────────────────────────────────────────

    /**
     * Downloads a single file from a redirect/tunnel URL with progress updates.
     * Mirrors {@code _download_file()} in Python.
     */
    private DownloaderService.DownloadResult downloadFile(
            String fileUrl,
            JsonNode metadata,
            Consumer<String> progressCallback,
            String originalUrl) throws IOException {

        // Sanitise filename
        String rawFilename = metadata.path("filename").asText(
                "video_" + UUID.randomUUID().toString().substring(0, 8) + ".mp4");
        String filename = rawFilename.replaceAll("[^a-zA-Z0-9._\\- ]", "");
        Path filePath = config.getDownloadsDir().resolve(filename);

        log.info("Downloading file from Cobalt: {}", fileUrl);

        Request request = new Request.Builder().url(fileUrl).build();
        try (Response resp = http.newCall(request).execute()) {
            if (!resp.isSuccessful() || resp.body() == null) {
                throw new IOException("Cobalt download failed: HTTP " + resp.code());
            }
            try (InputStream in = resp.body().byteStream();
                 OutputStream out = Files.newOutputStream(filePath)) {
                byte[] buf = new byte[8192];
                int n;
                while ((n = in.read(buf)) != -1) out.write(buf, 0, n);
            }
        }

        if (!Files.exists(filePath) || Files.size(filePath) == 0) {
            throw new IOException("Downloaded file is empty or missing: " + filePath);
        }
        log.info("File downloaded: {} ({} bytes)", filePath, Files.size(filePath));

        // Build metadata
        String uploader = firstNonBlank(
                metadata.path("author").asText(null),
                metadata.path("uploader").asText(null),
                metadata.path("creator").asText(null)
        );
        if (uploader == null && originalUrl != null) {
            Matcher m = Pattern.compile("tiktok\\.com/@([^/?\\s]+)").matcher(originalUrl);
            if (m.find()) uploader = m.group(1);
        }
        uploader = uploader != null ? uploader : "Unknown";

        String title = metadata.path("title").asText(null);
        if (title == null) title = filename.replaceFirst("\\.[^.]+$", "");

        Map<String, String> meta = new HashMap<>();
        meta.put("title", title);
        meta.put("uploader", uploader);
        meta.put("webpage_url", originalUrl != null ? originalUrl : "");
        meta.put("duration", "0");
        meta.put("verified", "false");
        meta.put("codec", "h264");

        // Enrich TikTok metadata via TikWM
        return new DownloaderService.DownloadResult(filePath, null, meta);
    }

    // ─── Picker (multi-file) ──────────────────────────────────────────────────────

    /**
     * Handles Cobalt picker response (e.g. TikTok slideshow).
     * Mirrors {@code _handle_picker()} in Python.
     */
    private DownloaderService.DownloadResult handlePicker(JsonNode response) throws IOException {
        log.info("Cobalt returned picker (multiple files)");

        JsonNode pickerItems = response.path("picker");
        if (!pickerItems.isArray() || pickerItems.isEmpty()) {
            throw new IOException("Cobalt picker response is empty");
        }

        List<Path> filePaths = new ArrayList<>();
        for (int i = 0; i < pickerItems.size(); i++) {
            JsonNode item = pickerItems.get(i);
            String itemUrl = item.path("url").asText(null);
            if (itemUrl == null) continue;

            String mime = item.path("mime").asText(item.path("type").asText(null));
            String ext = guessExtension(itemUrl, mime);
            String filename = "picker_" + i + ext;

            try {
                DownloaderService.DownloadResult r = downloadFile(
                        itemUrl,
                        mapper.createObjectNode().put("filename", filename),
                        null,
                        null);
                filePaths.add(r.getSingleFile());
            } catch (Exception e) {
                log.error("Failed to download picker item {}: {}", i, e.getMessage());
            }
        }

        if (filePaths.isEmpty()) throw new IOException("All picker items failed to download");

        String title = response.path("title").asText("Carousel Media");
        String uploader = firstNonBlank(
                response.path("author").asText(null),
                response.path("uploader").asText(null),
                response.path("creator").asText(null));

        Map<String, String> meta = new HashMap<>();
        meta.put("title", title);
        meta.put("uploader", uploader != null ? uploader : "Unknown");
        meta.put("webpage_url", response.path("url").asText(""));
        meta.put("count", String.valueOf(filePaths.size()));

        return new DownloaderService.DownloadResult(filePaths, null, meta);
    }

    // ─── Helpers ──────────────────────────────────────────────────────────────────

    private String guessExtension(String url, String mime) {
        if (url != null) {
            try {
                String path = new URI(url).getPath();
                int dot = path.lastIndexOf('.');
                if (dot >= 0) {
                    String ext = path.substring(dot).toLowerCase();
                    if (!ext.isBlank()) return ext;
                }
            } catch (Exception ignored) {}
        }
        if (mime != null) {
            String m = mime.toLowerCase();
            if (m.contains("jpeg")) return ".jpg";
            if (m.contains("png")) return ".png";
            if (m.contains("webp")) return ".webp";
            if (m.contains("mp4")) return ".mp4";
            if (m.contains("mp3")) return ".mp3";
            if (m.contains("m4a")) return ".m4a";
        }
        return ".bin";
    }

    private String firstNonBlank(String... values) {
        for (String v : values) {
            if (v != null && !v.isBlank()) return v;
        }
        return null;
    }
}
